React/CSS-in-JS/Chakra.md

```diff
@@ -0,0 +1,657 @@
+
+[All provider props](https://chakra-ui.com/docs/getting-started#chakraprovider-props)
+
+### Style props
+
+Style props are a way to alter the style of a component by simply passing props to it. It helps to save time by providing helpful shorthand ways to style components.
+Alternative to that is using chakra factory.
+
+Style props should be used for one-off styles (up to max 3 props):
+
+```tsx
+<Button display="block" mt="md">
+  Click me!
+</Button>
+```
+
+Components that have more than 3 styles should be moved to separate component.
+Every `Box` or `Flex` components should be moved to separate component, regarding of number of styles, with a descriptive name.
+
+(NOTE: Where to create that component will be added after we decide!)
+
+```tsx
+const ListWrapper = chakra(Box, {
+  baseStyle: {
+    shadow: "lg",
+    rounded: "lg",
+    bg: "white",
+  },
+});
+```
+
+For more about style props read [docs](https://chakra-ui.com/docs/features/style-props)
+
+`sx` prop works similar to `style` prop. It takes an object with styles.
+
+Example:
+
+```tsx
+<Button
+  sx={{
+    display: "block",
+    mt: "md",
+    color: "primary.100",
+    mx: 12,
+  }}
+>
+  Click me!
+</Button>
+```
+
+There is also `__css` prop that is similar to `sx`. The only difference is that `__css` prop will be merged before `sx` prop ([github](https://github.com/chakra-ui/chakra-ui/blob/085891be806b855af90b86367f5b26e8151c3ff5/packages/system/src/system.ts#L106)).
+
+### Chakra Factory
+
+Chakra factory serves as an object of chakra JSX elements, and also a function that can be used to enable custom component receive chakra's style props.
+
+```tsx
+import { chakra } from "@chakra-ui/react";
+
+// preferred way!
+chakra("button", {
+  baseStyle: {
+    shadow: "lg",
+    rounded: "lg",
+    bg: "white",
+  },
+});
+
+// using sx prop
+<chakra.button
+  sx={{
+    shadow: "lg",
+    rounded: "lg",
+    bg: "white",
+  }}
+>
+  Click me
+</chakra.button>;
+```
+
+This reduces the need to create custom component wrappers and name them. Syntax is available for common html elements. See the reference for the full [list of elements](https://github.com/chakra-ui/chakra-ui/blob/develop/packages/system/src/system.utils.ts#L9) supported.
+
+Chakra factory function should be only used in atom components. They will be defined with base style and can be overridden by `sx` prop because `baseStyle` is merged before `sx` ([merge order](https://github.com/chakra-ui/chakra-ui/blob/085891be806b855af90b86367f5b26e8151c3ff5/packages/system/src/system.ts#L104))
+
+#### Custom component example
+
+For example `react-datepicker` can be wrapped in chakra factory function so we can pass `sx` or style props to `<DatePicker />`
+
+Example:
+
+```tsx
+import DatePicker from "react-datepicker";
+import { chakra } from "@chakra-ui/react";
+
+const StyledaDatepicker = chakra(DatePicker);
+
+export const Datepicker = () => <StyledaDatepicker w="100%" bg="blue.100" />;
```

Comment:
Well, you're already using the chakra factory so might as well style it there.

```
const Datepicker = chakra(DatePicker, {
  baseStyle: {
    w: '100%', // or 'full'
    bg: 'blue.100',
  },
});
```

Reply:
It seems this demonstrates how to use theme-aware one-off styles with components that are not part of chakra.
A SImilar thing could be achieved with polymorphic `as` prop.
`<Box as="DatePicker" w="100%" bg="blue.100" />`
This is totally fine to me...